File: Core/announce_entry.cpp
#include "announce_entry.h"


#include "interop.h"
#include "error_code.h"

using namespace Tsunami::Core;

AnnounceEntry::AnnounceEntry(libtorrent::announce_entry& entry)
{
    entry_ = new libtorrent::announce_entry(entry);
}

AnnounceEntry::AnnounceEntry(System::String^ url)
{
    entry_ = new libtorrent::announce_entry(interop::to_std_string(url));
}

AnnounceEntry::~AnnounceEntry()
{
    delete entry_;
}

libtorrent::announce_entry* AnnounceEntry::ptr()
{
    return entry_;
}

int AnnounceEntry::next_announce_in()
{
    return entry_->next_announce_in();
}

int AnnounceEntry::min_announce_in()
{
    return entry_->min_announce_in();
}

void AnnounceEntry::reset()
{
    entry_->reset();
}

bool AnnounceEntry::can_announce(System::DateTime now, bool is_seed)
{
    System::DateTime unix(1970, 1, 1);
    int totalSeconds = System::Convert::ToInt32((now - unix.ToLocalTime()).TotalSeconds);
	const std::chrono::seconds sc(totalSeconds);
    return entry_->can_announce(libtorrent::ptime(sc), is_seed); // TODO: investigate ptime
}

bool AnnounceEntry::is_working()
{
    return entry_->is_working();
}

void AnnounceEntry::trim()
{
    entry_->trim();
}

System::String^ AnnounceEntry::url::get()
{
    return interop::from_std_string(entry_->url);
}

System::String^ AnnounceEntry::trackerid::get()
{
    return interop::from_std_string(entry_->trackerid);
}

System::String^ AnnounceEntry::message::get()
{
    return interop::from_std_string(entry_->message);
}

ErrorCode ^ AnnounceEntry::last_error::get()
{
	return gcnew ErrorCode(entry_->last_error);
}

System::DateTime AnnounceEntry::next_announce::get()
{
	std::chrono::milliseconds ms = std::chrono::duration_cast<std::chrono::milliseconds>(entry_->next_announce.time_since_epoch());
	return System::DateTime(1970, 1, 1, 0, 0, 0, System::DateTimeKind::Utc).AddMilliseconds(static_cast<double>(ms.count()));
}

System::DateTime AnnounceEntry::min_announce::get()
{
	std::chrono::milliseconds ms = std::chrono::duration_cast<std::chrono::milliseconds>(entry_->min_announce.time_since_epoch());
	return System::DateTime(1970, 1, 1, 0, 0, 0, System::DateTimeKind::Utc).AddMilliseconds(static_cast<double>(ms.count()));
}

int AnnounceEntry::scrape_incomplete::get()
{
    return entry_->scrape_incomplete;
}

int AnnounceEntry::scrape_complete::get()
{
    return entry_->scrape_complete;
}

int AnnounceEntry::scrape_downloaded::get()
{
    return entry_->scrape_downloaded;
}

int AnnounceEntry::tier::get()
{
    return entry_->tier;
}

int AnnounceEntry::fail_limit::get()
{
    return entry_->fail_limit;
}

int AnnounceEntry::fails::get()
{
    return entry_->fails;
}

bool AnnounceEntry::updating::get()
{
    return entry_->updating;
}

int AnnounceEntry::source::get()
{
    return entry_->source;
}

bool AnnounceEntry::verified::get()
{
    return entry_->verified;
}

bool AnnounceEntry::start_sent::get()
{
    return entry_->start_sent;
}

bool AnnounceEntry::complete_sent::get()
{
    return entry_->complete_sent;
}

bool AnnounceEntry::send_stats::get()
{
    return entry_->send_stats;
}

File: SessionManager/Models/AnnounceEntry.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tsunami.Models
{
    public class AnnounceEntry
    {
        public bool CompleteSent { get; set; }
        public int Fails { get; set; }
        public int FailLimit { get; set; }
        public ErrorCode LastError { get; set; }
        public string Message { get; set; }
        public DateTime MinAnnounce { get; set; }
        public DateTime NextAannounce { get; set; }
        public int ScrapeComplete { get; set; }
        public int ScrapeDownloaded { get; set; }
        public int ScrapeIncomplete { get; set; }
        public bool SendStats { get; set; }
        public int Source { get; set; }
        public bool StartSent { get; set; }
        public int Tier { get; set; }
        public string Trackerid { get; set; }
        public bool Updating { get; set; }
        public string Url { get; set; }
        public bool Verified { get; set; }
        public bool IsWorking { get; set; }
        public int MinAnnounceIn { get; set; }
        public int NextAnnounceIn { get; set; }

        public AnnounceEntry() { /* nothing to do. just for serializator */ }

        public AnnounceEntry(Core.AnnounceEntry ae)
        {
            CompleteSent = ae.complete_sent;
            Fails = ae.fails;
            FailLimit = ae.fail_limit;
            LastError = new ErrorCode(ae.last_error);
            Message = ae.message;
            MinAnnounce = ae.min_announce;
            NextAannounce = ae.next_announce;
            ScrapeComplete = ae.scrape_complete;
            ScrapeDownloaded = ae.scrape_downloaded;
            ScrapeIncomplete = ae.scrape_incomplete;
            SendStats = ae.send_stats;
            Source = ae.source;
            StartSent = ae.start_sent;
            Tier = ae.tier;
            Trackerid = ae.trackerid;
            Updating = ae.updating;
            Url = ae.url;
            Verified = ae.verified;
            IsWorking = ae.is_working();
            MinAnnounceIn = ae.min_announce_in();
            NextAnnounceIn = ae.next_announce_in();
        }

        //public void Reset()
        //{
        //    _ae.reset();
        //}

        //public void Trim()
        //{
        //    _ae.trim();
        //}

        //public bool CanAnnounce(DateTime now, bool is_seed)
        //{
        //    return _ae.can_announce(now, is_seed);
        //}

    }
}
